feat!: Refactor

Dronakurl opened the pull request:
It is such a great project to learn rust.
I refactored it to allow any kind of modules as long as the implement Display.
I am not mad when you decline this PR, it was a great learning about polymorphisms.

BTW: I don't think, the threads are necessary: tmux already runs commands in the background and everything but the CPU stuff is really fast (1ms). But it might speed things up for future modules

Dronakurl committed:
feat(docs): add detailed documentation and refactor code

- Added documentation comments for `Color` enum and `Style` struct in
`colors/mod.rs`
- Documented `Display` implementation for `Style` and `Color`
- Added documentation for `foreground_color`, `background_color`, and
`bold` functions
- Updated `config.rs` to use new modules and styles
- Enhanced `icons` module with detailed documentation
- Refactored `main.rs` to remove unnecessary threading
- Updated `README.md` with detailed instructions for adding custom
modules
- Added `battery.rs` module for battery status
- Enhanced `datetime.rs` module with date and time functions
- Refactored `systemstats.rs` to use `fmt::Display` instead of `Show`
- Improved `tmux.rs` with detailed documentation and standard styling
- Removed unused `conditional_insert.rs` and `system` utilities
- Updated `utils/strings.rs` with detailed documentation
- Updated `modules/README.md` with contributing guidelines
- Refactored `modules/mod.rs` to include new modules

Dronakurl committed:
chore: Update crate for systemstats

Dronakurl committed:
feat: add NVIDIA module

Dronakurl committed:
feat: percentage display for nvidia module

Dronakurl committed:
Change default config

Dronakurl committed:
doc: improve README

Dronakurl committed:
feat: high CPU module

Dronakurl committed:
backup high cpu

Dronakurl committed:
finish high cpu module

Dlurak commented:
I'm gonna review it in a few days at the earliest, thank you so much for your efforts :)

As you've been adding new commits to the pr it would be good to know when you think that it is ready from your site, feel free to ping me once you are happy with the pr so I know that reviewing makes sense :)

Dronakurl committed:
Update NVIDIA module

Dronakurl committed:
fix: handle no battery

Dronakurl committed:
feat: Make highcpu colors configurable

Dronakurl committed:
feat: color options more configurable in module standards

Dronakurl committed:
fix: empty modules

Dronakurl committed:
doc: add clarification in README

Dronakurl committed:
Merge branch 'master' into refactor

Dronakurl commented:
> I'm gonna review it in a few days at the earliest, thank you so much for your efforts :)
> 
> As you've been adding new commits to the pr it would be good to know when you think that it is ready from your site, feel free to ping me once you are happy with the pr so I know that reviewing makes sense :)

I think, I am happy with the result now

Dlurak commented on src/colors/mod.rs:
I know that I already did it like that but can you please move that file to `src/color.rs`. It's kind of unnecessary to have a new directory for only one file

Dlurak commented on src/icons/mod.rs:
Same as for the `colors/mod.rs` it would be better to move the file to `src/icons.rs`

Dlurak commented on src/utils/mod.rs:
This entire `utils` module now only offers the `PrettyDuration` struct. I think we can delete the directory and move `src/utils/string.rs` to something like `src/pretty_duration.rs`

Dlurak commented on src/modules/high_cpu.rs:
```diff
@@ -0,0 +1,94 @@
+            None => return Box::new(Module::default()),
+        };
+
+        let process_info = if output.status.success() {
+            let stdout = String::from_utf8_lossy(&output.stdout);
+            let mut process_map: HashMap<String, f32> = HashMap::new();
+
+            stdout.lines().skip(1).for_each(|line| {
+                let parts: Vec<&str> = line.split_whitespace().collect();
+                if let (Some(cpu_usage), Some(name)) = (parts.get(1), parts.get(2)) {
+                    let cpu_usage: f32 = cpu_usage.parse().unwrap_or(0.0);
+                    if cpu_usage > 0.001 && *name != "ps" {
+                        *process_map.entry(name.to_string()).or_insert(0.0) += cpu_usage;
+                    }
+                }
+            });
```
Rust has methods on Iterators and Options that allow for creating very short, declerative and (imho) easy to read code:
```rust
let process_map = stdout
    .lines()
    .skip(1)
    .filter_map(|line| {
        let mut parts = line.split_whitespace().skip(1);
        let cpu_usage: f32 = parts.next()?.parse().unwrap_or(0.0);
        let name = parts.next()?;
        (cpu_usage > 0.1 && name != "ps").then_some((name, cpu_usage))
    })
    .fold(HashMap::new(), |mut acc, (name, cpu_usage)| {
        *acc.entry(name).or_insert(0.0) += cpu_usage;
        acc
    });
```

Btw, the type of the HashMap does not need to be explicitly stated, as the line which adds/updates entries uses `f32` and `String` the type can be inferred

Dlurak commented on src/icons/mod.rs:
```diff
@@ -1,30 +1,61 @@
 pub enum Icon {
+    /// Default empty icon
+    #[default]
+    Empty,
```
We use `Option<Icon>` everywhere. It seems like `None` and `Icon::Empty` represent the same thing. I would prefer to remove that `Empty` variant. Do you have a good reason to keep it?

Dlurak commented on .archive/nvidia.txt:
What reason does this file serve, can it be removed? Can the entire `.archive` directory be removed?

Dlurak commented on src/colors/mod.rs:
```diff
@@ -11,63 +16,100 @@ pub enum Color {
     Cyan,
     Magenta,
     Reset,
+    Any(&'static str),
```
This is a great idea! Can you please document that variant that it either can be a tmux color code like `colour165` or a hex color code like `#ff56af`.

Dlurak commented on src/modules/README.md:
````diff
@@ -1,14 +1,69 @@
+        }
+    }
+
+    impl CustomModule {
+        pub fn new() -> Box<Module<CustomModule>> {
+            Box::new(Module::new(
+                CustomModule,
+                Some(Icon::Custom),
+                Style {
+                    fg: Color::Blue,
+                    bg: Color::Reset,
+                    bold: true,
+                },
+            ))
+        }
+    }
+    ```
+
+2. Add your module to the `get_modules` function in `src/config.rs`:
+
+    ```rust
+    use crate::modules::custom::CustomModule;
+
+    pub fn get_modules() -> Vec<Box<dyn Display + Send>> {
+        vec![
+            // Existing modules...
+            CustomModule::new(),
````
When you implement the new proposed trait also update this example please.

Dlurak commented on src/modules/battery.rs:
```diff
@@ -0,0 +1,145 @@
+                is_charging,
+            })
+        } else {
+            Ok(Self {
+                percentages: 0,
+                is_charging: false,
+            })
```
I think a `Ok(None)` would be the best option here. In rust it is generally discouraged to use magic values like 0 to indicate something. In this case you use the special value of 0 to indicate that no battery exists, this would be better with a `Option`. Also consider removing/modifying the `Default` implementation
When changing this you would of course also need to modify all usages of that special value for example in the `Display` implementation.

Dlurak commented on src/modules/battery.rs:
```diff
@@ -0,0 +1,145 @@
+    ///
+    /// # Returns
+    /// A boxed Module containing battery information with conditional styling
+    pub fn get_with_warning(bg: Option<Color>) -> Box<Module<String>> {
```
There is no advantage in coupling the warning to the battery, it makes it harder to style, impossible to use independently and limits how the two modules can be placed in the status bar. 
For example someone might want to place a green battery on the far left and a red battery warning on the far right, that would need two independent modules.
Not too important but it would be great if the threshold weren't hardcoded to twenty but a parameter

Dlurak commented on src/modules/high_cpu.rs:
```diff
@@ -0,0 +1,94 @@
+    /// # Returns
+    ///
+    /// A `Box<Module<HighCpuModule>>` containing the process information.
+    pub fn new(fg: Option<Color>, bg: Option<Color>) -> Box<Module<HighCpu>> {
```
Instead of having the errors as the process info this should return a `Result`. Then the caller can decide how to handle the errors, maybe they want to show an error text with a different style, maybe they want to ignore the error...

Dlurak commented on src/modules/mod.rs:
```diff
@@ -1,79 +1,61 @@
-                let swap_usage_percent = (used_swap as f64 / total_swap as f64) * 100.0;
-                Ok(strings::round(swap_usage_percent, rounding))
-            }
+impl<T> Default for Module<T>
```
When T implements the new proposed trait the style should be read from the trait implementation.

Btw I love your idea with the new Module struct, great job :+1:

Dlurak commented on src/modules/nvidia.rs:
```diff
@@ -0,0 +1,78 @@
+use std::str;
+
+pub struct Nvidia {
+    pub memory_used: String,
+    pub memory_total: String,
```
A Option of some number type should be better here. I don't have a nvidia card so I can't say which type but some number type

Dlurak reviewed (changes requested):
Sorry that it took me so long to review your pr, I worked on some things and then forgot about it, I'm really sorry.
Wow that is really some quite impressive work! :+1:
But there are a few general small issues that need optimization:

---

Can you please create a new traits to convert the actual modules like `Memory` or `Nvidia` to the `Module` struct.
Something like that might be a good:

```rust
trait ModuleStyles {
    fn style(&self) -> Style;
    fn icon(&self) -> Option<Icon>;
}
```

`Module` should get a method to create a new method from something that implements `ModuleStyles`. `Module::default()` should also use the new trait when possible.

---

Many modules instantly return a `Box<Module<Self>>`.
To make them more versatile they could return `Self` or in some cases `Option<Self>`
Using the new trait it's easy to create a `Box<Module>` from it, just something like one of the two lines would be enough:

```rust
Module::new_box(Memory::default());
Box::new(Module::default_styled(Memory::default()));
```

---

It's ok when a module is optional or can fail, for example the low battery warning should be this: `Option<Module<String>>`.
`get_modules()` could use the `.chain` method for that.

---

Also I left you some comments on the code which need to be addressed :)

Dlurak commented:
I've just refactored various parts myself.

Dronakurl commented:
Thank you for looking into it! I started too, but I did not immediately get
the idea with the new trait.
Can I still help with this?

Dlurak ***@***.***> schrieb am So. 23. Feb. 2025 um 21:14:

> I've just refactored various parts myself.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Dlurak/muxbar/pull/4#issuecomment-2677091991>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASKZU4B42DPWD52GYWYXC6T2RITY5AVCNFSM6AAAAABUL4PWZSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNZXGA4TCOJZGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> [image: Dlurak]*Dlurak* left a comment (Dlurak/muxbar#4)
> <https://github.com/Dlurak/muxbar/pull/4#issuecomment-2677091991>
>
> I've just refactored various parts myself.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Dlurak/muxbar/pull/4#issuecomment-2677091991>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASKZU4B42DPWD52GYWYXC6T2RITY5AVCNFSM6AAAAABUL4PWZSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNZXGA4TCOJZGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Dlurak commented:
If you want to you can port your new modules and documentation. Totally up to you :)

---

I've fundamentally changed how muxbar renders it's modules:

Previously tmux would run muxbar, wait a short moment and then run it again... After every fresh start muxbar had to recompute every single module.

Now each module specifies when it will be rerendered and also how to update itself.
Once a module needs to rerender muxbar rerenders the desired module while using all other modules from a cache.
Muxbar now runs indefinitely and just outputs new lines from time to time. 
For this reason the `ToModule` trait got two new methods.

Dronakurl commented:
I am no longer using tmux, so I close this. Good luck!
